File: week10/5/prob3.py
class Solution:
    def exist(self, board: list[list[str]], word: str) -> bool:
        vertices, adj = self.board_to_adj(board)
        result = False
        for source in vertices:
            result = result or self.dfs_helper(source, vertices, adj, 0, word, [])
            if result:
                break
        return result

    def dfs_helper(self, curr: int, vertices: dict[int, str], adj: dict[int, list[int]], level: int, word: str, visited: list[int]):
        # print(f"[Level {level} in {word}], curr : {vertices[curr]}({curr})")
        # base case
        # - 불일치
        if vertices[curr] != word[level]:
            # print("BACKTRACKING!")
            return False
        # - 일치 & 끝까지 도달
        elif level == len(word) - 1:
            # print(f"[{level}] {curr} : {vertices[curr]} / {word[level]}")
            return True

        # recursive case
        # - 일치 & 아직 끝까지 도달하지 못함
        else:
            result = False
            for neighbor in adj[curr]:
                if neighbor not in visited:
                    this_result = self.dfs_helper(neighbor, vertices, adj, level+1, word, visited+[curr])
                    # if this_result:
                    #     print(f"[{level}] {neighbor} : {vertices[neighbor]} / {word[level]}")
                    # print()
                    result = result or this_result
                if result: break
            return result

    def board_to_adj(self, board: list[list[str]]) -> tuple[dict[int, str], dict[int, list[int]]]:
        directions = [
                    [-1, 0],
            [0, -1],         [0, 1],
                    [1, 0]
        ]

        vertices = dict()
        adj = dict()

        M = len(board)
        N = len(board[0])

        for i in range(M):
            for j in range(N):
                node = i * N + j
                vertices[node] = board[i][j]
                adj[node] = []
                for d in directions:
                    ii = i + d[0]
                    jj = j + d[1]
                    if 0 <= ii < M and 0 <= jj < N:
                        neighbor = ii * N + jj
                        adj[node].append(neighbor)
        
        return vertices, adj

File: week12/5/prob3.cpp
#include <vector>
#include <unordered_map>
using namespace std;

class Solution {
public:
    int countBattleships(vector<vector<char>>& board) {
        int numIslands = 0;
        unordered_map<int, vector<int>> adj = convertGridToAdj(board);
        unordered_map<int, bool> visited;
        for(const auto& pair : adj) {
            visited[pair.first] = false;
        }
        for(const auto& pair : adj) {
            if (!visited[pair.first]) {
                dfs_from_source(pair.first, adj, visited);
                numIslands += 1;
            }
        }
        
        return numIslands;
    }

    unordered_map<int, vector<int>> convertGridToAdj(vector<vector<char>>& grid) {
        int M = grid.size();
        int N = grid[0].size();
        vector<vector<int>> directions = {
                        {-1, 0},
                {0, -1},        {0, 1},
                        { 1, 0}
        };

        unordered_map<int, vector<int>> adj;
        for (int i=0; i<M; i++) {
            for (int j=0; j<N; j++) {
                if (grid[i][j] == 'X') {
                    int curr = i * N + j;
                    adj[curr] = {};
                    for (vector<int> d : directions) {
                        int ii = i + d[0];
                        int jj = j + d[1];
                        if (ii>=0 && ii<M && jj>=0 && jj<N && grid[ii][jj] == 'X') {
                            int neighbor = ii * N + jj;
                            adj[curr].push_back(neighbor);
                        }
                    }
                }
            }
        }
        return adj;
    }

    void dfs_from_source(int source, unordered_map<int, vector<int>>& adj, unordered_map<int, bool>& visited) {
        visited[source] = true;
        for (int neighbor : adj[source]) {
            if(!visited[neighbor]) {
                dfs_from_source(neighbor, adj, visited);
            }
        }
    }
};
